File: android_benchmark/app/src/main/java/com/nanogpt/benchmark/MainActivity.java
package com.nanogpt.benchmark;

import android.app.Activity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends Activity {
    static { System.loadLibrary("nanogpt_benchmark"); }

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private TextView output;

    private native String runBenchmark(int promptTokens, int generateTokens, int threads);

    @Override public void onCreate(Bundle state) {
        super.onCreate(state);
        output = new TextView(this);
        output.setTextSize(14f);
        output.setPadding(24, 24, 24, 24);
        output.setText("NanoGPT Android Benchmark\n\n" +
                "Convert a checkpoint to app/src/main/assets/nanogpt.bin, build, install, then tap Run.\n" +
                "This JNI path executes a simple fp32 CPU forward pass for repeatable hardware comparisons.\n");

        Button run = new Button(this);
        run.setText("Run 32-token benchmark");
        run.setOnClickListener(v -> startBenchmark());

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.addView(run, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        ScrollView scroller = new ScrollView(this);
        scroller.addView(output);
        root.addView(scroller, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f));
        setContentView(root);
    }

    private void startBenchmark() {
        output.append("\nRunning...\n");
        executor.submit(() -> {
            long start = System.nanoTime();
            String result = runBenchmark(16, 32, Runtime.getRuntime().availableProcessors());
            double wall = (System.nanoTime() - start) / 1_000_000_000.0;
            runOnUiThread(() -> output.append(String.format(Locale.US, "%s\nWall time: %.3fs\n", result, wall)));
        });
    }

    @Override protected void onDestroy() {
        executor.shutdownNow();
        super.onDestroy();
    }
}

File: android_app/app/src/main/java/com/nanogpt/benchmark/MainActivity.kt
package com.nanogpt.benchmark

import android.app.Activity
import android.os.Bundle
import android.text.InputType
import android.view.Gravity
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import java.util.Locale
import kotlin.concurrent.thread

class MainActivity : Activity() {
    private lateinit var promptInput: EditText
    private lateinit var maxTokensInput: EditText
    private lateinit var runButton: Button
    private lateinit var resultPanel: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        buildUi()
    }

    private fun buildUi() {
        val density = resources.displayMetrics.density
        fun dp(value: Int): Int = (value * density).toInt()

        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(20), dp(20), dp(20), dp(20))
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
        }

        root.addView(TextView(this).apply {
            text = "nanoGPT Android Benchmark"
            textSize = 24f
            gravity = Gravity.CENTER_HORIZONTAL
        })

        promptInput = EditText(this).apply {
            hint = "Prompt"
            minLines = 3
            setText("To be, or not to be")
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE
        }
        root.addView(promptInput)

        maxTokensInput = EditText(this).apply {
            hint = "max_new_tokens"
            setText("32")
            inputType = InputType.TYPE_CLASS_NUMBER
        }
        root.addView(maxTokensInput)

        runButton = Button(this).apply {
            text = "Run benchmark"
            setOnClickListener { runBenchmark() }
        }
        root.addView(runButton)

        resultPanel = TextView(this).apply {
            textSize = 16f
            setPadding(0, dp(16), 0, 0)
            text = startupMessage()
        }
        root.addView(resultPanel)

        setContentView(ScrollView(this).apply { addView(root) })
    }

    private fun runBenchmark() {
        val prompt = promptInput.text.toString()
        val maxNewTokens = maxTokensInput.text.toString().toIntOrNull()?.coerceAtLeast(1) ?: 32
        runButton.isEnabled = false
        resultPanel.text = "Running warmup and measured generation..."

        thread(name = "nanogpt-benchmark") {
            val output = try {
                createModelRunner().use { runner ->
                    val result = BenchmarkRunner(runner).run(prompt, maxNewTokens)
                    formatResult(result)
                }
            } catch (t: Throwable) {
                "Benchmark failed:\n${t.message}\n\n${startupMessage()}"
            }

            runOnUiThread {
                resultPanel.text = output
                runButton.isEnabled = true
            }
        }
    }

    private fun createModelRunner(): ModelRunner {
        val assets = assets.list("")?.toSet().orEmpty()
        return when {
            "model.onnx" in assets -> OrtModelRunner(this)
            "model.pte" in assets -> ExecuTorchModelRunner(this)
            else -> error("Add model.onnx or model.pte to app/src/main/assets before running.")
        }
    }

    private fun formatResult(result: GenerationResult): String = buildString {
        appendLine("Generated text:")
        appendLine(result.generatedText)
        appendLine()
        appendLine("Prompt token count: ${result.promptTokenCount}")
        appendLine("Generated token count: ${result.generatedTokenCount}")
        appendMetric("TTFT ms", result.ttftMs)
        appendMetric("TPOT ms/token", result.tpotMs)
        appendMetric("Total generation time ms", result.totalGenerationMs)
        appendMetric("tokens/sec", result.tokensPerSecond)
        appendMetric("Java/Kotlin heap MB", result.javaHeapMb)
        appendMetric("native heap MB", result.nativeHeapMb)
        appendMetric("Android PSS MB", result.androidPssMb)
    }

    private fun StringBuilder.appendMetric(label: String, value: Double) {
        appendLine("$label: ${String.format(Locale.US, "%.2f", value)}")
    }

    private fun startupMessage(): String = """
        Copy an exported model into app/src/main/assets before running:
        - model.onnx for ONNX Runtime Mobile
        - model.pte for the ExecuTorch placeholder path

        Also replace meta.json with the tokenizer metadata that matches the exported nanoGPT checkpoint.
    """.trimIndent()
}
